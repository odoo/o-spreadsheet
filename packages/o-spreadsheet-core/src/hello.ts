export function helloWorldFromCorePackage() {
  console.log("Hello, World!");
}
