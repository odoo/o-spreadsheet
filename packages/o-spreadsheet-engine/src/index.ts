export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
