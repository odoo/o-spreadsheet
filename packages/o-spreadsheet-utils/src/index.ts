export * from "./translation";
export * from "./uuid";
