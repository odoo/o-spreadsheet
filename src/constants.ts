import { _t } from "./translation";
import { BorderDescr, Color, Style } from "./types";
import { CellErrorType } from "./types/errors";

export const CANVAS_SHIFT = 0.5;

// Colors
export const BACKGROUND_GRAY_COLOR = "#f5f5f5";
export const BACKGROUND_HEADER_COLOR = "#F8F9FA";
export const BACKGROUND_HEADER_SELECTED_COLOR = "#E8EAED";
export const BACKGROUND_HEADER_ACTIVE_COLOR = "#595959";
export const TEXT_HEADER_COLOR = "#666666";
export const FIGURE_BORDER_COLOR = "#c9ccd2";
export const SELECTION_BORDER_COLOR = "#3266ca";
export const HEADER_BORDER_COLOR = "#C0C0C0";
export const CELL_BORDER_COLOR = "#E2E3E3";
export const BACKGROUND_CHART_COLOR = "#FFFFFF";
export const BG_HOVER_COLOR = "#EBEBEB";
export const DISABLED_TEXT_COLOR = "#CACACA";
export const DEFAULT_COLOR_SCALE_MIDPOINT_COLOR = 0xb6d7a8;
export const LINK_COLOR = "#01666b";
export const FILTERS_COLOR = "#188038";
export const BACKGROUND_HEADER_FILTER_COLOR = "#E6F4EA";
export const BACKGROUND_HEADER_SELECTED_FILTER_COLOR = "#CEEAD6";
export const SEPARATOR_COLOR = "#E0E2E4";
export const ICONS_COLOR = "#4A4F59";
export const HEADER_GROUPING_BACKGROUND_COLOR = "#F5F5F5";
export const HEADER_GROUPING_BORDER_COLOR = "#999";
export const GRID_BORDER_COLOR = "#E2E3E3";
export const FROZEN_PANE_HEADER_BORDER_COLOR = "#BCBCBC";
export const FROZEN_PANE_BORDER_COLOR = "#DADFE8";
export const COMPOSER_ASSISTANT_COLOR = "#9B359B";
export const HIGHLIGHT_COLOR = "#FE019A";

// Color picker defaults as upper case HEX to match `toHex`helper
export const COLOR_PICKER_DEFAULTS: Color[] = [
  "#000000",
  "#434343",
  "#666666",
  "#999999",
  "#B7B7B7",
  "#CCCCCC",
  "#D9D9D9",
  "#EFEFEF",
  "#F3F3F3",
  "#FFFFFF",
  "#980000",
  "#FF0000",
  "#FF9900",
  "#FFFF00",
  "#00FF00",
  "#00FFFF",
  "#4A86E8",
  "#0000FF",
  "#9900FF",
  "#FF00FF",
  "#E6B8AF",
  "#F4CCCC",
  "#FCE5CD",
  "#FFF2CC",
  "#D9EAD3",
  "#D0E0E3",
  "#C9DAF8",
  "#CFE2F3",
  "#D9D2E9",
  "#EAD1DC",
  "#DD7E6B",
  "#EA9999",
  "#F9CB9C",
  "#FFE599",
  "#B6D7A8",
  "#A2C4C9",
  "#A4C2F4",
  "#9FC5E8",
  "#B4A7D6",
  "#D5A6BD",
  "#CC4125",
  "#E06666",
  "#F6B26B",
  "#FFD966",
  "#93C47D",
  "#76A5AF",
  "#6D9EEB",
  "#6FA8DC",
  "#8E7CC3",
  "#C27BA0",
  "#A61C00",
  "#CC0000",
  "#E69138",
  "#F1C232",
  "#6AA84F",
  "#45818E",
  "#3C78D8",
  "#3D85C6",
  "#674EA7",
  "#A64D79",
  "#85200C",
  "#990000",
  "#B45F06",
  "#BF9000",
  "#38761D",
  "#134F5C",
  "#1155CC",
  "#0B5394",
  "#351C75",
  "#741B47",
  "#5B0F00",
  "#660000",
  "#783F04",
  "#7F6000",
  "#274E13",
  "#0C343D",
  "#1C4587",
  "#073763",
  "#20124D",
  "#4C1130",
];

// Dimensions
export const MIN_ROW_HEIGHT = 10;
export const MIN_COL_WIDTH = 5;
export const HEADER_HEIGHT = 26;
export const HEADER_WIDTH = 48;
export const TOPBAR_HEIGHT = 63;
export const TOPBAR_TOOLBAR_HEIGHT = 34;
export const BOTTOMBAR_HEIGHT = 36;
export const DEFAULT_CELL_WIDTH = 96;
export const DEFAULT_CELL_HEIGHT = 23;
export const SCROLLBAR_WIDTH = 15;
export const AUTOFILL_EDGE_LENGTH = 8;
export const ICON_EDGE_LENGTH = 18;
export const UNHIDE_ICON_EDGE_LENGTH = 14;
export const MIN_CF_ICON_MARGIN = 4;
export const MIN_CELL_TEXT_MARGIN = 4;
export const CF_ICON_EDGE_LENGTH = 15;
export const PADDING_AUTORESIZE_VERTICAL = 3;
export const PADDING_AUTORESIZE_HORIZONTAL = MIN_CELL_TEXT_MARGIN;
export const GROUP_LAYER_WIDTH = 21;
export const GRID_ICON_MARGIN = 2;
export const GRID_ICON_EDGE_LENGTH = 17;
export const FILTER_ICON_MARGIN = 2;
export const FILTER_ICON_EDGE_LENGTH = 17;
export const FOOTER_HEIGHT = 2 * DEFAULT_CELL_HEIGHT;

// Menus
export const MENU_WIDTH = 250;
export const MENU_VERTICAL_PADDING = 6;
export const MENU_ITEM_HEIGHT = 26;
export const MENU_ITEM_PADDING_HORIZONTAL = 11;
export const MENU_ITEM_PADDING_VERTICAL = 4;
export const MENU_SEPARATOR_BORDER_WIDTH = 1;
export const MENU_SEPARATOR_PADDING = 5;
export const MENU_SEPARATOR_HEIGHT = MENU_SEPARATOR_BORDER_WIDTH + 2 * MENU_SEPARATOR_PADDING;

// Style
export const DEFAULT_STYLE = {
  align: "left",
  verticalAlign: "bottom",
  wrapping: "overflow",
  bold: false,
  italic: false,
  strikethrough: false,
  underline: false,
  fontSize: 10,
  fillColor: "",
  textColor: "#000000",
} satisfies Required<Style>;

export const DEFAULT_VERTICAL_ALIGN = DEFAULT_STYLE.verticalAlign;
export const DEFAULT_WRAPPING_MODE = DEFAULT_STYLE.wrapping;

// Fonts
export const DEFAULT_FONT_WEIGHT = "400";
export const DEFAULT_FONT_SIZE = DEFAULT_STYLE.fontSize;
export const HEADER_FONT_SIZE = 11;
export const DEFAULT_FONT = "'Roboto', arial";

// Borders
export const DEFAULT_BORDER_DESC: BorderDescr = { style: "thin", color: "#000000" };
export const DEFAULT_FILTER_BORDER_DESC: BorderDescr = { style: "thin", color: FILTERS_COLOR };

// Ranges
export const INCORRECT_RANGE_STRING = CellErrorType.InvalidReference;

// Max Number of history steps kept in memory
export const MAX_HISTORY_STEPS = 99;

// Id of the first revision
export const DEFAULT_REVISION_ID = "START_REVISION";

// Figure
export const DEFAULT_FIGURE_HEIGHT = 335;
export const DEFAULT_FIGURE_WIDTH = 536;
export const FIGURE_BORDER_WIDTH = 1;
export const ACTIVE_BORDER_WIDTH = 2;
export const MIN_FIG_SIZE = 80;

// Chart
export const MAX_CHAR_LABEL = 20;
export const FIGURE_ID_SPLITTER = "??";

export const DEFAULT_GAUGE_LOWER_COLOR = "#cc0000";
export const DEFAULT_GAUGE_MIDDLE_COLOR = "#f1c232";
export const DEFAULT_GAUGE_UPPER_COLOR = "#6aa84f";

export const DEFAULT_SCORECARD_BASELINE_MODE = "difference";
export const DEFAULT_SCORECARD_BASELINE_COLOR_UP = "#00A04A";
export const DEFAULT_SCORECARD_BASELINE_COLOR_DOWN = "#DC6965";

export const LINE_FILL_TRANSPARENCY = 0.4;

// session
export const DEBOUNCE_TIME = 200;
export const MESSAGE_VERSION = 1;

// Sheets
export const FORBIDDEN_SHEET_CHARS = ["'", "*", "?", "/", "\\", "[", "]"] as const;
export const FORBIDDEN_IN_EXCEL_REGEX = /'|\*|\?|\/|\\|\[|\]/;

// Cells
export const FORMULA_REF_IDENTIFIER = "|";
export const LOADING = "Loading...";
export const DEFAULT_ERROR_MESSAGE = _t("Invalid expression");

// Components
export enum ComponentsImportance {
  Grid = 0,
  Highlight = 5,
  HeaderGroupingButton = 6,
  Figure = 10,
  ScrollBar = 15,
  GridPopover = 19,
  GridComposer = 20,
  Dropdown = 21,
  IconPicker = 25,
  TopBarComposer = 30,
  Popover = 35,
  FigureAnchor = 1000,
  FigureSnapLine = 1001,
}
let DEFAULT_SHEETVIEW_SIZE = 0;

export function getDefaultSheetViewSize() {
  return DEFAULT_SHEETVIEW_SIZE;
}

export function setDefaultSheetViewSize(size: number) {
  DEFAULT_SHEETVIEW_SIZE = size;
}

export const MAXIMAL_FREEZABLE_RATIO = 0.85;

export const NEWLINE = "\n";
export const FONT_SIZES: number[] = [6, 7, 8, 9, 10, 11, 12, 14, 18, 24, 36];
