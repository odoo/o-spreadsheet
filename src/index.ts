import { ChartFigure } from "./components/figures/figure_chart/figure_chart";
import { BasicChartPanel } from "./components/side_panel/chart/basic_chart_panel/basic_chart_panel";
import { ScorecardChartPanel } from "./components/side_panel/chart/scorecard_chart_panel/scorecard_chart_panel";
import {
  BOTTOMBAR_HEIGHT,
  DEFAULT_CELL_HEIGHT,
  DEFAULT_CELL_WIDTH,
  HEADER_HEIGHT,
  HEADER_WIDTH,
  MIN_COL_WIDTH,
  MIN_ROW_HEIGHT,
  SCROLLBAR_WIDTH,
  TOPBAR_HEIGHT,
} from "./constants";
import { toBoolean, toJsDate, toNumber, toString } from "./functions/helpers";
import { args, functionRegistry } from "./functions/index";
import { LinkCell } from "./helpers/cells/index";
import {
  computeTextWidth,
  formatValue,
  isMarkdownLink,
  markdownLink,
  numberToLetters,
  parseMarkdownLink,
  toCartesian,
  toXC,
  toZone,
  UuidGenerator,
} from "./helpers/index";
import { createEmptyWorkbookData } from "./migrations/data";
import { corePluginRegistry, uiPluginRegistry } from "./plugins/index";
import {
  autofillModifiersRegistry,
  autofillRulesRegistry,
  cellMenuRegistry,
  cellRegistry,
  colMenuRegistry,
  createFullMenuItem,
  dashboardMenuRegistry,
  inverseCommandRegistry,
  linkMenuRegistry,
  otRegistry,
  rowMenuRegistry,
  sheetMenuRegistry,
  sidePanelRegistry,
  topbarComponentRegistry,
  topbarMenuRegistry,
} from "./registries/index";

/**
 * We export here all entities that needs to be accessed publicly by Odoo.
 *
 * Note that the __info__ key is actually completed by the build process (see
 * the rollup.config.js file)
 */

export const __info__ = {};
export { Revision } from "./collaborative/revisions";
export { Spreadsheet } from "./components/index";
export { DATETIME_FORMAT } from "./constants";
export { compile, functionCache } from "./formulas/compiler";
export { astToFormula, convertAstNodes, parse } from "./formulas/parser";
export { Model } from "./model";
export { CorePlugin } from "./plugins/core_plugin";
export { UIPlugin } from "./plugins/ui_plugin";
export { Registry } from "./registry";
export { setTranslationMethod } from "./translation";
export { CancelledReason, CommandResult, DispatchResult } from "./types";
export { Client } from "./types/collaborative/session";
export {
  ClientJoinedMessage,
  ClientLeftMessage,
  ClientMovedMessage,
  CollaborationMessage,
  RemoteRevisionMessage,
  RevisionRedoneMessage,
  RevisionUndoneMessage,
  TransportService,
} from "./types/collaborative/transport_service";
export { coreTypes, invalidateEvaluationCommands, readonlyAllowedCommands } from "./types/commands";
export const SPREADSHEET_DIMENSIONS = {
  MIN_ROW_HEIGHT,
  MIN_COL_WIDTH,
  HEADER_HEIGHT,
  HEADER_WIDTH,
  TOPBAR_HEIGHT,
  BOTTOMBAR_HEIGHT,
  DEFAULT_CELL_WIDTH,
  DEFAULT_CELL_HEIGHT,
  SCROLLBAR_WIDTH,
};

export const registries = {
  autofillModifiersRegistry,
  autofillRulesRegistry,
  cellMenuRegistry,
  colMenuRegistry,
  dashboardMenuRegistry,
  linkMenuRegistry,
  functionRegistry,
  uiPluginRegistry,
  corePluginRegistry,
  rowMenuRegistry,
  sidePanelRegistry,
  sheetMenuRegistry,
  topbarMenuRegistry,
  topbarComponentRegistry,
  otRegistry,
  inverseCommandRegistry,
  cellRegistry,
};
export const cellTypes = {
  LinkCell,
};
export const helpers = {
  args,
  toBoolean,
  toNumber,
  toString,
  toXC,
  toJsDate,
  toZone,
  toCartesian,
  numberToLetters,
  createFullMenuItem,
  UuidGenerator,
  formatValue,
  computeTextWidth,
  isMarkdownLink,
  parseMarkdownLink,
  markdownLink,
  createEmptyWorkbookData,
};
export const components = {
  ChartFigure,
  BasicChartPanel,
  ScorecardChartPanel,
};
