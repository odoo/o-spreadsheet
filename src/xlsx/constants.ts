import { ExcelIconSet } from "../types/xlsx";

export const XLSX_FORMAT_MAP = {
  General: 0,
  "0": 1,
  "0.00": 2,
  "#,#00": 3,
  "#,##0.00": 4,
  "0%": 9,
  "0.00%": 10,
  "0.00E+00": 11,
  "# ?/?": 12,
  "# ??/??": 13,
  "mm-dd-yy": 14,
  "d-mm-yy": 15,
  "mm-yy": 16,
  "mmm-yy": 17,
  "h:mm AM/PM": 18,
  "h:mm:ss AM/PM": 19,
  "h:mm": 20,
  "h:mm:ss": 21,
  "m/d/yy h:mm": 22,
  "#,##0 ;(#,##0)": 37,
  "#,##0 ;[Red](#,##0)": 38,
  "#,##0.00;(#,##0.00)": 39,
  "#,##0.00;[Red](#,##0.00)": 40,
  "mm:ss": 45,
  "[h]:mm:ss": 46,
  "mmss.0": 47,
  "##0.0E+0": 48,
  "@": 49,
  "hh:mm:ss a": 19, // TODO: discuss: this format is not recognized by excel for example (doesn't follow their guidelines I guess)
};

export const XLSX_ICONSET_MAP: Record<string, ExcelIconSet> = {
  arrow: "3Arrows",
  smiley: "3Symbols",
  dot: "3TrafficLights1",
};

export const NAMESPACE = {
  styleSheet: "http://schemas.openxmlformats.org/spreadsheetml/2006/main",
  sst: "http://schemas.openxmlformats.org/spreadsheetml/2006/main",
  Relationships: "http://schemas.openxmlformats.org/package/2006/relationships",
  Types: "http://schemas.openxmlformats.org/package/2006/content-types",
  worksheet: "http://schemas.openxmlformats.org/spreadsheetml/2006/main",
  workbook: "http://schemas.openxmlformats.org/spreadsheetml/2006/main",
  drawing: "http://schemas.openxmlformats.org/drawingml/2006/spreadsheetDrawing",
};

export const DRAWING_NS_A = "http://schemas.openxmlformats.org/drawingml/2006/main";
export const DRAWING_NS_C = "http://schemas.openxmlformats.org/drawingml/2006/chart";

export const CONTENT_TYPES = {
  workbook: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml",
  sheet: "application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml",
  sharedStrings: "application/vnd.openxmlformats-officedocument.spreadsheetml.sharedStrings+xml",
  styles: "application/vnd.openxmlformats-officedocument.spreadsheetml.styles+xml",
  drawing: "application/vnd.openxmlformats-officedocument.drawing+xml",
  chart: "application/vnd.openxmlformats-officedocument.drawingml.chart+xml",
} as const;

export const RELATIONSHIP_NSR =
  "http://schemas.openxmlformats.org/officeDocument/2006/relationships";

export const HEIGHT_FACTOR = 0.75; // 100px => 75 u
export const WIDTH_FACTOR = 0.1425; // 100px => 14.25 u

export const FIRST_NUMFMT_ID = 164;

export interface functionDefaultArg {
  type: "NUMBER";
  value: number;
}

export const FORCE_DEFAULT_ARGS_FUNCTIONS: Record<string, functionDefaultArg[]> = {
  FLOOR: [{ type: "NUMBER", value: 1 }],
  CEILING: [{ type: "NUMBER", value: 1 }],
  ROUND: [{ type: "NUMBER", value: 0 }],
  ROUNDUP: [{ type: "NUMBER", value: 0 }],
  ROUNDDOWN: [{ type: "NUMBER", value: 0 }],
};

/**
 * This list contains all "future" functions that are not compatible with older versions of Excel
 * For more information, see https://docs.microsoft.com/en-us/openspecs/office_standards/ms-xlsx/5d1b6d44-6fc1-4ecd-8fef-0b27406cc2bf
 */
export const NON_RETROCOMPATIBLE_FUNCTIONS = [
  "ACOT",
  "ACOTH",
  "AGGREGATE",
  "ARABIC",
  "BASE",
  "BETA.DIST",
  "BETA.INV",
  "BINOM.DIST",
  "BINOM.DIST.RANGE",
  "BINOM.INV",
  "BITAND",
  "BITLSHIFT",
  "BITOR",
  "BITRSHIFT",
  "BITXOR",
  "CEILING.MATH",
  "CEILING.PRECISE",
  "CHISQ.DIST",
  "CHISQ.DIST.RT",
  "CHISQ.INV",
  "CHISQ.INV.RT",
  "CHISQ.TEST",
  "COMBINA",
  "CONCAT",
  "CONFIDENCE.NORM",
  "CONFIDENCE.T",
  "COT",
  "COTH",
  "COVARIANCE.P",
  "COVARIANCE.S",
  "CSC",
  "CSCH",
  "DAYS",
  "DECIMAL",
  "ERF.PRECISE",
  "ERFC.PRECISE",
  "EXPON.DIST",
  "F.DIST",
  "F.DIST.RT",
  "F.INV",
  "F.INV.RT",
  "F.TEST",
  "FILTERXML",
  "FLOOR.MATH",
  "FLOOR.PRECISE",
  "FORECAST.ETS",
  "FORECAST.ETS.CONFINT",
  "FORECAST.ETS.SEASONALITY",
  "FORECAST.ETS.STAT",
  "FORECAST.LINEAR",
  "FORMULATEXT",
  "GAMMA",
  "GAMMA.DIST",
  "GAMMA.INV",
  "GAMMALN.PRECISE",
  "GAUSS",
  "HYPGEOM.DIST",
  "IFNA",
  "IFS",
  "IMCOSH",
  "IMCOT",
  "IMCSC",
  "IMCSCH",
  "IMSEC",
  "IMSECH",
  "IMSINH",
  "IMTAN",
  "ISFORMULA",
  "ISOWEEKNUM",
  "LOGNORM.DIST",
  "LOGNORM.INV",
  "MAXIFS",
  "MINIFS",
  "MODE.MULT",
  "MODE.SNGL",
  "MUNIT",
  "NEGBINOM.DIST",
  "NORM.DIST",
  "NORM.INV",
  "NORM.S.DIST",
  "NORM.S.INV",
  "NUMBERVALUE",
  "PDURATION",
  "PERCENTILE.EXC",
  "PERCENTILE.INC",
  "PERCENTRANK.EXC",
  "PERCENTRANK.INC",
  "PERMUTATIONA",
  "PHI",
  "POISSON.DIST",
  "QUARTILE.EXC",
  "QUARTILE.INC",
  "QUERYSTRING",
  "RANK.AVG",
  "RANK.EQ",
  "RRI",
  "SEC",
  "SECH",
  "SHEET",
  "SHEETS",
  "SKEW.P",
  "STDEV.P",
  "STDEV.S",
  "SWITCH",
  "T.DIST",
  "T.DIST.2T",
  "T.DIST.RT",
  "T.INV",
  "T.INV.2T",
  "T.TEST",
  "TEXTJOIN",
  "UNICHAR",
  "UNICODE",
  "VAR.P",
  "VAR.S",
  "WEBSERVICE",
  "WEIBULL.DIST",
  "XOR",
  "Z.TEST",
];
