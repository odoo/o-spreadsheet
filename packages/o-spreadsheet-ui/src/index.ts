import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}