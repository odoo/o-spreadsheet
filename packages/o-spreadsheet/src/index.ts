import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };