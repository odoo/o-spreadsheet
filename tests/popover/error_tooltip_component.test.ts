import { Model } from "../../src";
import {
  ErrorToolTip,
  ErrorToolTipMessage,
} from "../../src/components/error_tooltip/error_tooltip";
import { DEFAULT_CELL_HEIGHT, DEFAULT_CELL_WIDTH } from "../../src/constants";
import {
  addDataValidation,
  createChart,
  merge,
  setCellContent,
} from "../test_helpers/commands_helpers";
import { TEST_CHART_DATA } from "../test_helpers/constants";
import {
  clickCell,
  gridMouseEvent,
  hoverCell,
  triggerMouseEvent,
  triggerWheelEvent,
} from "../test_helpers/dom_helper";
import {
  makeTestComposerStore,
  mockChart,
  mountComponent,
  mountSpreadsheet,
  nextTick,
} from "../test_helpers/helpers";

mockChart();

describe("Error tooltip component", () => {
  let fixture: HTMLElement;

  async function mountErrorTooltip(errors: ErrorToolTipMessage[]) {
    ({ fixture } = await mountComponent(ErrorToolTip, { props: { errors } }));
  }

  test("Can display an error message", async () => {
    await mountErrorTooltip([{ message: "This is an error", title: "Error" }]);
    expect(fixture.querySelector(".o-error-tooltip-title")?.textContent).toBe("Error");
    expect(fixture.querySelector(".o-error-tooltip-message")?.textContent).toBe("This is an error");
  });

  test("Can display multiple error messages", async () => {
    await mountErrorTooltip([
      { message: "This is an error", title: "Error" },
      { message: "Invalid data", title: "Invalid" },
    ]);
    const titles = fixture.querySelectorAll(".o-error-tooltip-title");
    const messages = fixture.querySelectorAll(".o-error-tooltip-message");
    expect(titles).toHaveLength(2);
    expect(messages).toHaveLength(2);

    expect(titles[0].textContent).toBe("Error");
    expect(messages[0].textContent).toBe("This is an error");

    expect(titles[1].textContent).toBe("Invalid");
    expect(messages[1].textContent).toBe("Invalid data");
  });
});

describe("Grid integration", () => {
  let model: Model;
  let fixture: HTMLElement;

  beforeEach(async () => {
    jest.useFakeTimers();
    ({ model, fixture } = await mountSpreadsheet());
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test("can display error on A1", async () => {
    setCellContent(model, "A1", "=1/0");
    await hoverCell(model, "A1", 400);
    expect(document.querySelector(".o-error-tooltip")).not.toBeNull();
  });

  test("can display invalid data validation error", async () => {
    setCellContent(model, "A1", "hello");
    addDataValidation(model, "A1", "id", { type: "textContains", values: ["hi"] });
    await hoverCell(model, "A1", 400);
    expect(document.querySelector(".o-error-tooltip")).not.toBeNull();
  });

  test("can display both cell error and data validation error", async () => {
    setCellContent(model, "A1", "=1/0");
    addDataValidation(model, "A1", "id", { type: "textContains", values: ["1"] });
    await hoverCell(model, "A1", 400);
    expect(document.querySelectorAll(".o-error-tooltip-title")).toHaveLength(2);
  });

  test("don't display error on #N/A", async () => {
    Date.now = jest.fn(() => 0);
    setCellContent(model, "A1", "=NA()");
    await nextTick();
    gridMouseEvent(model, "pointermove", "A1");
    Date.now = jest.fn(() => 500);
    jest.advanceTimersByTime(300);
    await nextTick();
    expect(document.querySelector(".o-error-tooltip")).toBeNull();
  });

  test("Display error on #N/A 'non-silent' ", async () => {
    Date.now = jest.fn(() => 0);
    setCellContent(model, "A1", "=VLOOKUP(6,A1:A2,B2:B4)");
    await nextTick();
    gridMouseEvent(model, "pointermove", "A1");
    Date.now = jest.fn(() => 500);
    jest.advanceTimersByTime(300);
    await nextTick();
    expect(document.querySelector(".o-error-tooltip")).not.toBeNull();
  });

  test("can display error tooltip", async () => {
    setCellContent(model, "C8", "=1/0");
    await hoverCell(model, "C8", 200);
    expect(document.querySelector(".o-error-tooltip")).toBeNull();
    await hoverCell(model, "C8", 400);
    expect(document.querySelector(".o-error-tooltip")).not.toBeNull();
    expect(document.querySelector(".o-error-tooltip")?.parentElement).toMatchSnapshot();
    await hoverCell(model, "C7", 200);
    expect(document.querySelector(".o-error-tooltip")).not.toBeNull();
    await hoverCell(model, "C7", 400);
    expect(document.querySelector(".o-error-tooltip")).toBeNull();
  });

  test("can display error when move on merge", async () => {
    merge(model, "C1:C8");
    setCellContent(model, "C1", "=1/0");
    await hoverCell(model, "C8", 400);
    expect(document.querySelector(".o-error-tooltip")).not.toBeNull();
  });

  test("Hovering over a figure should not open popovers", async () => {
    createChart(model, { ...TEST_CHART_DATA.basicChart }, "figureId");
    model.dispatch("UPDATE_FIGURE", {
      id: "figureId",
      y: 200,
      x: 200,
      width: 200,
      height: 200,
      sheetId: model.getters.getActiveSheetId(),
    });
    await nextTick();
    setCellContent(model, "C3", "[label](url.com)");

    triggerMouseEvent(".o-figure", "pointermove", DEFAULT_CELL_WIDTH * 2, DEFAULT_CELL_HEIGHT * 2);
    jest.advanceTimersByTime(400);
    await nextTick();

    expect(fixture.querySelector(".o-popover")).toBeNull();
  });

  test("composer content is set when clicking on merged cell (not top left)", async () => {
    const composerStore = makeTestComposerStore(model);
    merge(model, "C1:C8");
    setCellContent(model, "C1", "Hello");
    await nextTick();
    await clickCell(model, "C8");
    expect(composerStore.currentContent).toBe("Hello");
  });

  test("Wheel events on error tooltip are scrolling the grid", async () => {
    setCellContent(model, "C1", "=0/0");
    await hoverCell(model, "C1", 400);
    triggerWheelEvent(".o-error-tooltip", { deltaY: 300, deltaX: 300 });
    await nextTick();
    expect(fixture.querySelector(".o-scrollbar.vertical")?.scrollTop).toBe(300);
    expect(fixture.querySelector(".o-scrollbar.horizontal")?.scrollLeft).toBe(300);
  });
});
