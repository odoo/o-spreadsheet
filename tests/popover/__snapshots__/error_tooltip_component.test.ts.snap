// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid integration can display error tooltip 1`] = `
<div
  class="o-popover"
  style="z-index: 19; display: block; max-height: 824px; max-width: 697px; top: 187px; left: 336px;"
>
  <div
    class="o-error-tooltip"
  >
    <div
      class="o-error-tooltip-title fw-bold text-danger"
    >
      Error
    </div>
    <div
      class="o-error-tooltip-message"
    >
      The divisor must be different from zero.
    </div>
    
  </div>
</div>
`;
